rust/worker/src/quorum_waiter.rs

```diff
@@ -0,0 +1,86 @@
+use crate::processor::SerializedBatchMessage;
+use config::{Committee, Stake};
+use crypto::PublicKey;
+use futures::stream::futures_unordered::FuturesUnordered;
+use futures::stream::StreamExt as _;
+use network::CancelHandler;
+use tokio::sync::mpsc::{Receiver, Sender};
+
+#[cfg(test)]
+#[path = "tests/quorum_waiter_tests.rs"]
+pub mod quorum_waiter_tests;
+
+#[derive(Debug)]
+pub struct QuorumWaiterMessage {
+    /// A serialized `WorkerMessage::Batch` message.
+    pub batch: SerializedBatchMessage,
+    /// The cancel handlers to receive the acknowledgements of our broadcast.
+    pub handlers: Vec<(PublicKey, CancelHandler)>,
+}
+
+/// The QuorumWaiter waits for 2f authorities to acknowledge reception of a batch.
+pub struct QuorumWaiter {
```

Comment:
I assume there will be another QuorumWaiter for the primary messages? Is there a reason this is not part of the network and we specialize it for batches?

Reply:
The QuorumWaiter is worker-specific because it is sequential: we handle one batch at the time. We are not broadcasting the next batch until we receive enough ACK from the previous one. The primary needs something similar (called CertificateWaiter and HeaderWaiter) that can wait for more messages at the same time